File: projecteuler.fsharp/CommonLib/PandigitalNumbers.fs
namespace CommonLib

open System

module PandigitalNumbersDef =

    [<Literal>]
    let Pandigital2Min = 12L

    [<Literal>]
    let Pandigital2Max = 21L

    [<Literal>]
    let Pandigital3Min = 123L

    [<Literal>]
    let Pandigital3Max = 321L

    [<Literal>]
    let Pandigital4Min = 1234L

    [<Literal>]
    let Pandigital4Max = 4321L

    [<Literal>]
    let Pandigital5Min = 12345L

    [<Literal>]
    let Pandigital5Max = 54321L

    [<Literal>]
    let Pandigital6Min = 123456L

    [<Literal>]
    let Pandigital6Max = 654321L

    [<Literal>]
    let Pandigital7Min = 1234567L

    [<Literal>]
    let Pandigital7Max = 7654321L

    [<Literal>]
    let Pandigital8Min = 12345678L

    [<Literal>]
    let Pandigital8Max = 87654321L

    [<Literal>]
    let Pandigital9Min = 123456789L

    [<Literal>]
    let Pandigital9Max = 987654321L

    [<Literal>]
    let Pandigital10Min = 1023456789L

    [<Literal>]
    let Pandigital10Max = 9876543210L

[<AbstractClass; Sealed>]
type PandigitalNumbers =

    static member public IsPandigital(number: int) =
        number |> int64 |> PandigitalNumbers.IsPandigital

    static member public IsPandigital(number: int64) =
        let rec isPandigital (number: int64) (expectedValue: int64) (actualValue: int64) =
            match number with
            | 0L when expectedValue = actualValue -> true
            | 0L when expectedValue <> actualValue -> false
            | _ -> isPandigital (number / 10L) expectedValue (actualValue ||| (1L <<< ((number % 10L) |> int)))
        match number with
        | _ when number < 0L -> raise (ArgumentOutOfRangeException("number"))
        | _ when (PandigitalNumbersDef.Pandigital2Min <= number) && (number <= PandigitalNumbersDef.Pandigital2Max) -> isPandigital number 0b0000000110L 0L
        | _ when (PandigitalNumbersDef.Pandigital3Min <= number) && (number <= PandigitalNumbersDef.Pandigital3Max) -> isPandigital number 0b0000001110L 0L
        | _ when (PandigitalNumbersDef.Pandigital4Min <= number) && (number <= PandigitalNumbersDef.Pandigital4Max) -> isPandigital number 0b0000011110L 0L
        | _ when (PandigitalNumbersDef.Pandigital5Min <= number) && (number <= PandigitalNumbersDef.Pandigital5Max) -> isPandigital number 0b0000111110L 0L
        | _ when (PandigitalNumbersDef.Pandigital6Min <= number) && (number <= PandigitalNumbersDef.Pandigital6Max) -> isPandigital number 0b0001111110L 0L
        | _ when (PandigitalNumbersDef.Pandigital7Min <= number) && (number <= PandigitalNumbersDef.Pandigital7Max) -> isPandigital number 0b0011111110L 0L
        | _ when (PandigitalNumbersDef.Pandigital8Min <= number) && (number <= PandigitalNumbersDef.Pandigital8Max) -> isPandigital number 0b0111111110L 0L
        | _ when (PandigitalNumbersDef.Pandigital9Min <= number) && (number <= PandigitalNumbersDef.Pandigital9Max) -> isPandigital number 0b1111111110L 0L
        | _ when (PandigitalNumbersDef.Pandigital10Min <= number) && (number <= PandigitalNumbersDef.Pandigital10Max) -> isPandigital number 0b1111111111L 0L
        | _ -> false

    static member public IsPandigital(number: bigint) =
        number |> int64 |> PandigitalNumbers.IsPandigital

File: projecteuler.erlang/common/src/pandigital_numbers.erl
%% @author std-string

-module(pandigital_numbers).
-export([is_pandigital/1]).

-include("pandigital_def.hrl").

%% ====================================================================
%% API functions
%% ====================================================================

-spec is_pandigital(Number :: pos_integer()) -> boolean().
is_pandigital(Number) ->
    if
        (?PANDIGITAL2_MIN =< Number) and (Number =< ?PANDIGITAL2_MAX) -> is_pandigital_impl(Number, 0, 2#110);
        (?PANDIGITAL3_MIN =< Number) and (Number =< ?PANDIGITAL3_MAX) -> is_pandigital_impl(Number, 0, 2#1110);
        (?PANDIGITAL4_MIN =< Number) and (Number =< ?PANDIGITAL4_MAX) -> is_pandigital_impl(Number, 0, 2#11110);
        (?PANDIGITAL5_MIN =< Number) and (Number =< ?PANDIGITAL5_MAX) -> is_pandigital_impl(Number, 0, 2#111110);
        (?PANDIGITAL6_MIN =< Number) and (Number =< ?PANDIGITAL6_MAX) -> is_pandigital_impl(Number, 0, 2#1111110);
        (?PANDIGITAL7_MIN =< Number) and (Number =< ?PANDIGITAL7_MAX) -> is_pandigital_impl(Number, 0, 2#11111110);
        (?PANDIGITAL8_MIN =< Number) and (Number =< ?PANDIGITAL8_MAX) -> is_pandigital_impl(Number, 0, 2#111111110);
        (?PANDIGITAL9_MIN =< Number) and (Number =< ?PANDIGITAL9_MAX) -> is_pandigital_impl(Number, 0, 2#1111111110);
        (?PANDIGITAL10_MIN =< Number) and (Number =< ?PANDIGITAL10_MAX) -> is_pandigital_impl(Number, 0, 2#1111111111);
        true -> false
    end.

%% ====================================================================
%% Internal functions
%% ====================================================================

-spec is_pandigital_impl(Number :: non_neg_integer(), Result :: non_neg_integer(), ExpectedValue :: pos_integer()) -> boolean().
is_pandigital_impl(0, ExpectedValue, ExpectedValue) -> true;
is_pandigital_impl(0, _Result, _ExpectedValue) -> false;
is_pandigital_impl(Number, Result, ExpectedValue) ->
    is_pandigital_impl(Number div 10, Result bor (1 bsl (Number rem 10)), ExpectedValue).